vcx/libvcx/src/messages/payload.rs

```diff
@@ -90,36 +133,57 @@ impl Payloads {
         }
     }
 
-    pub fn decrypt(my_vk: &str, payload: &MessagePayload) -> VcxResult<(String, Option<Thread>)> {
+    pub fn decrypt_helper(my_vk: &str, payload: &MessagePayload) -> VcxResult<(String, Option<Thread>, Option<MessageTypeV2>)> {
         match payload {
             MessagePayload::V1(payload) => {
+                // If we can do v1 style decryption
                 if let Ok(payload) = Payloads::decrypt_payload_v1(my_vk, payload) {
-                    Ok((payload.msg, None))
+                    // Return a VcxResult that contains the output String and no thread.
+                    Ok((payload.msg, None, None))
                 } else {
+                    // Convert to a vector of u8.
                     let vec = to_u8(payload);
+                    // Get a JSON value from the text -- or return VcxError on failure.
                     let json: Value = serde_json::from_slice(&vec[..])
-                        .map_err(|err| VcxError::from_msg(VcxErrorKind::InvalidMessagePack, format!("Cannot deserialize MessagePayload: {}", err)))?;
-
-                    let payload = match Payloads::decrypt_payload_v12(&my_vk, &json)?.msg {
+                        .map_err(|err|
+                            VcxError::from_msg(VcxErrorKind::InvalidMessagePack,
+                                               format!("Cannot deserialize MessagePayload: {}", err)))?;
+                    // If we got a JSON value, decrypt it in v12 style. If that fails, return
+                    // VcxError. If it succeeds, get String output from the .msg property. It
+                    // might be encoded as a String already, or it might be a JSON Value object
+                    // that needs to be converted to a String.
+                    let payload = Payloads::decrypt_payload_v12(&my_vk, &json)?;
+                    let type_ = payload.type_;
+                    let payload = match payload.msg {
                         serde_json::Value::String(_str) => _str,
                         value => value.to_string()
                     };
-
-                    Ok((payload, None))
+                    // Return a VcxResult that contains the output String and no thread.
+                    Ok((payload, None, Some(type_)))
                 }
             }
+            // Else if we can do v2 style decryption
             MessagePayload::V2(payload) => {
                 let payload = Payloads::decrypt_payload_v2(my_vk, payload)?;
-                Ok((payload.msg, Some(payload.thread)))
+                Ok((payload.msg, Some(payload.thread), None))
             }
         }
     }
 
+    pub fn decrypt(my_vk: &str, payload: &MessagePayload) -> VcxResult<(String, Option<Thread>)> {
+        match Payloads::decrypt_helper(my_vk, payload) {
+            Ok((data, th, _)) => Ok((data, th)),
+            Err(x) => Err(x)
+        }
```

Comment:
```suggestion
        Payloads::decrypt_helper(my_vk, payload)
            .map(|(data, th, _)| (data, th))
```

Reply:
I haven't try to compile this suggestion but it should be about to correct

Reply:
Thank you. I tried to write it this way yesterday, but I wasn't able to get it to compile. What I notice is different is that you put parens around the capture tuple. I had missed that. Thanks for helping me get it right.

Now updating.